Updating to current jenkins plugin behaviour, updated documentation

jdoose opened the pull request:
* Added parameter `user`, `trace`
* ignoring phase COMPLETED, but reacting on STARTED and FINALIZED. Removed phase FINISHED, does not exist anymore
* Added loggging
* Update documentation (issue #6 )

halkeye commented on src/scripts/hubot-jenkins-notifier.coffee:
```diff
@@ -91,17 +96,31 @@ class JenkinsNotifier
+
+    if query.room
+      console.log "sending to room #{query.room}"
+      envelope.user.room = envelope.room = query.room if query.room
```
Its already wrapped in query.room, so the if isn't ncessary

Can you change the console.log so its optional somehow?

halkeye commented on src/scripts/hubot-jenkins-notifier.coffee:
```diff
@@ -91,17 +96,31 @@ class JenkinsNotifier
+      return
+
+    if query.user
+      console.log "sending to user #{query.user}"
```
Can you change the console.log so its optional somehow?

halkeye commented on src/scripts/hubot-jenkins-notifier.coffee:
```diff
@@ -117,22 +136,46 @@ class JenkinsNotifier
     fullurl = data.build.full_url || data.build.url
 
-    if data.build.phase in ['FINISHED', 'FINALIZED']
+    if data.build.phase in ['COMPLETED']
```
Why were the older two removed?

jdoose replied:
The FINIHSED phase does not exist according to toe notification plugin. Valid settings are "Job Started", "Job Completed", "Job Finalized" or "All Events", they map to STARTED, COMPLETED and FINALIZED.
The STARTED is now also handled and sends a message to the room, if wanted.

I wanted the user to be able to leave (the default option, which ist "All events".
In that case COMPLETED and FINALIZED are sent. If both are handled then the mechanism to remember the last result was broken. 
So I decided to not handle COMPLETED at all.

So, first COMPLETED notifications are ignored, leaving STARTED and FINALIZED.
Then first STARTED is handled, after that the rest of the notifications.

But you are right, it could be handled better and more clearer. I will adjust the code.

Do you think it is necessary to still handle FINISHED since it is not supported in the plugin anymore?

jdoose commented:
Gavin, I updated the pull request as you recommended. I hope it is in a way you like it ;)
If not: please be patient, I will not be able to respond or upload anything new within in the next 10 days or so.

halkeye commented:
hey @jdoose its looking pretty good.
I can see your reasoning for removing the older states. Since apparently I didn't add tests for them, I think i knew at the time they weren't long for this code. I've found a lot of jenkins users don't regularly upgrade, so you have to keep backwards compatibility for a while.

I haven't had a chance to review this, and I probably won't till next weekend (I have a conference go to this week), For your new behavior, is the current testing good enough? Or do you want new test cases.

To me, if there's a test available, that means that functionality is going to stick around (thats why i'm okay deleting the older states)

halkeye commented:
So i found some uncommitted code locally
```
+#   Optional Params:
+#     Notification Strategy is [Ff][Ss] which stands for "Failure" and "Success"
+#     Capitalized letter means: notify always
+#       small letter means: notify only if buildstatus has changed
+#       "Fs" is the default
+#
```

Not sure if its worth mentioning any of this stuff in your new patch

halkeye commented on src/scripts/hubot-jenkins-notifier.coffee:
```diff
@@ -117,22 +140,50 @@ class JenkinsNotifier
+    if data.build.phase in ['STARTED']
+      @logMessage(query.trace, "#{data.name} #{data.build.phase} #{data.build.status}")
+
+      @robot.send envelope, "#{data.name} build ##{data.build.number} started: #{fullurl}"
```
this should be wrapped with `if @shouldNotify(envelope.notstrat, data)`

jdoose replied:
Hm, I am not so sure about this... As far as I understand `shouldNotify` is used to determine a state change. But for STARTED there is no state change, right? So I figured, if somebody turns on STARTED they probably want to be notified in any case.
But if you want to we can still wrap it in `shouldNotify`

halkeye replied:
Yes please

I'll be home Sunday and try to.give better feedback.

On Aug 19, 2016 1:49 PM, "Jens Doose" <notifications@github.com> wrote:

> In src/scripts/hubot-jenkins-notifier.coffee
> <https://github.com/halkeye/hubot-jenkins-notifier/pull/7#discussion_r75549311>
> :
>
> >      fullurl = data.build.full_url || data.build.url
> >
> > +    if data.build.phase in ['COMPLETED']
> > +      @logMessage(query.trace, "Ignoring phase COMPLETED")
> > +      return
> > +
> > +    if data.build.phase in ['STARTED']
> > +      @logMessage(query.trace, "#{data.name} #{data.build.phase} #{data.build.status}")
> > +
> > +      @robot.send envelope, "#{data.name} build ##{data.build.number} started: #{fullurl}"
>
> Hm, I am not so sure about this... As far as I understand shouldNotify is
> used to determine a state change. But for STARTED there is no state change,
> right? So I figured, if somebody turns on STARTED they probably want to be
> notified in any case.
> But if you want to we can still wrap it in shouldNotify
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/halkeye/hubot-jenkins-notifier/pull/7/files/a83dad2397b1f8afeee252d7df10018e88d38781#r75549311>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGuB5C6lK_Xk-B9T7MnC6AFViL7_bhwks5qhhbigaJpZM4JY0mC>
> .
>

halkeye commented on src/scripts/hubot-jenkins-notifier.coffee:
```diff
@@ -91,17 +100,31 @@ class JenkinsNotifier
     envelope.user.type = query.type if query.type
 
+    if query.user && query.room
+      @logMessage(query.trace, "Cannot use room and user together")
```
This doesn't make sense to be wrapped in logMessage. Its always an error, so always echo out the error.

I'm thinking it'd be worth also outputting both query.user and query.room so it can be debugged later.

jdoose replied:
Done and done.

jdoose commented:
I added the requested changes and added new test cases for the new errors with the parameters "room" and "user"

I am not so sure about the test cases, I added more for the error cases when both user and room parameter are given.
But I think the SUCCESS status test cases are missing. Plus, the last two test cases have phase "Started" and status "Failed"/"Finalized", is that even possible?

jdoose commented:
@halkeye , I am sorry but I was not able to get the tests running on GitHub with the specification of grunt-simple-mocha to ~0.4.0. There have been some recent changes in mocha and, quite frankly, I am too new to CoffeeScript to get it working.
It works now with pinning the version to 0.4.0, maybe you can take a look at it if you want it t be changed. Sorry about that!

halkeye commented:
Yea no problem.

I hate coffeescript, so I've learned enough to avoid it :D
Coffeescript automatically returns the last line of a function, in this case its the request function.
Mocha got updated to handle promises being returned so you don't need the done() callback.
The easiest solution is just to add a return at the end of your function

Take a look at https://github.com/halkeye/hubot-jenkins-notifier/commit/d65e17ccd266d68a20bc009623dab721fcedd9ca

Also should probably merge in/rebase master, i've been doing a bit of cleanup to get some deps up to date/running on travis.

jdoose commented:
Thanks for the hint, that did the trick! I merge in your master and updated the pull request.

halkeye commented:
Okay first pass looks good. I'll review when I'm back from the conference, so likely Sunday.

Thanks for being patent

halkeye commented on src/scripts/hubot-jenkins-notifier.coffee:
```diff
@@ -115,24 +144,51 @@ class JenkinsNotifier
+    @logMessage(query.trace, "#{data.name} #{data.build.phase} #{data.build.status}")
+
+    if data.build.phase in ['STARTED']
+      @robot.send envelope, "#{data.name} build ##{data.build.number} started: #{fullurl}"
```
I feel like I asked this before. But can you wrap this in shouldNotify? That should mean you won't have to change the test/old behavior.

jdoose replied:
;) Yes you did.
I wasn't so sure about this, since shouldNotify respects the current and the previous build outcome. But STARTED does not have an outcome, right?

halkeye replied:
Some people only want to know whats happening with a build when something
is broken.

Started will be based on the last build's outcome.

notStrat=
s = only notify on start if last job was a failure
S = always notify on start

On Mon, Aug 22, 2016 at 7:10 AM, Jens Doose <notifications@github.com>
wrote:

> In src/scripts/hubot-jenkins-notifier.coffee
> <https://github.com/halkeye/hubot-jenkins-notifier/pull/7#discussion_r75684234>
> :
>
> >      fullurl = data.build.full_url || data.build.url
> >
> > +    if data.build.phase in ['COMPLETED']
> > +      @logMessage(query.trace, "Ignoring phase COMPLETED")
> > +      return
> > +
> > +    @logMessage(query.trace, "#{data.name} #{data.build.phase} #{data.build.status}")
> > +
> > +    if data.build.phase in ['STARTED']
> > +      @robot.send envelope, "#{data.name} build ##{data.build.number} started: #{fullurl}"
>
> ;) Yes you did.
> I wasn't so sure about this, since shouldNotify respects the current and
> the previous build outcome. But STARTED does not have an outcome, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/halkeye/hubot-jenkins-notifier/pull/7/files/ebbd9282c7440d21e2e2fdb107daf33d10d2e09a#r75684234>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGuB4pasc0kjE4wjP2FsWNFEp1gOFomks5qia3ygaJpZM4JY0mC>
> .
>

halkeye commented:
was standalone-endpoint-server meant to be committed?

jdoose commented:
Hello halkeye, yes, it was intended to be committed.
I needed a test server to receive notifications from the jenkins plugin to check what was really sent over the wire. I like to check in test programs that might be handy later as well. If you don't agree it can be deleted, no problem.
Is there any way to exclude the code from codacy tests since it is only a test program and not a part of the hubot plugin? And do you know a way to run the codacy checks locally before pushing

halkeye commented:
All the codacy errors I can see are ones reported by eslint.

Please remove the extra test programs. They don't belong in this module.

Also very confused what the new logic is addressing. All I asked for was a single line of code to be added. Now there is a lot of new logic. I just really wanted to make sure the original functionality was kept, which the missing line and the changed tests broke.

halkeye commented:
Oh, looking back at the first commit now

You are adding support for messaging a user directly
Also for notify a job starts
Adding build log url

Plus a slew of new tests
Tons of new debugging logging.

I thought the original bug was just that it didn't handle finalized stage

jdoose commented:
I will remove the test program.

It did not work with the currently existing implementation since `shouldNotify` works on `data.build.status`, which is *not* included in a STARTED notification.
So I needed an extra handling in `shouldNotify` to check for phase STARTED and also check for the previous test result, as you requested. To avoid doubled code I introduced the new functions `lastBuildFailed`/`lastBuildSucceeded`
New tests were necessary because the STARTED notification should result in a message when a previous build failed. That was not possible with the previously existing tests.
So in my understanding I implemented it because you requested the call to shouldNotify. Which I thought was the better way, so I agreed.

halkeye commented:
Okay I'll look at it again tomorrow. I only just realized the SF is success
and failure and not start and finish.a

On Aug 31, 2016 12:09 AM, "Jens Doose" <notifications@github.com> wrote:

> I will remove the test program.
>
> It did not work with the currently existing implementation since
> shouldNotify works on data.build.status, which is *not* included in a
> STARTED notification.
> So I needed an extra handling in shouldNotify to check for phase STARTED
> and also check for the previous test result, as you requested. To avoid
> doubled code I introduced the new functions lastBuildFailed/lastBuildSucce
> eded
> New tests were necessary because the STARTED notification should result in
> a message when a previous build failed. That was not possible with the
> previously existing tests.
> So in my understanding I implemented it because you requested the call to
> shouldNotify. Which I thought was the better way, so I agreed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/halkeye/hubot-jenkins-notifier/pull/7#issuecomment-243677691>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGuBymZTXLeurgzTAtx86zLXQXFnTSxks5qlSijgaJpZM4JY0mC>
> .
>

halkeye commented:
Okay, I havn't had enough time to look at it i'm sorry. That'll change tomorrow. I was distracted by another project.

I think functionality wise we are totally there, I just don't want to see the defaults change.
I'm thinking we warn and drop the old query string values, and make some new ones.
Kinda tempted to just merge and make those changes, but I was thinking

`?onStart=SF&onFinished=sF&onWhatever=sf`

then `?nostat=SF` = `?onStart=&onFinished=SF`

type thing

What do you think? Clearer, configurable, without breaking existing behavior.
I'll also see if i can't make the tests a little easier to write, the big arrays are frustrating.

jdoose commented:
I like the idea. That really looks good! 
Do you want to merge it and implement the changes? Unfortunately I won't have much time in the next few days... :/

halkeye commented:
Yep. I'll merge, get the new configs in, and do a release.

Thanks for all your hard work.

halkeye commented:
@jdoose okay, Big rewrite today, removed coffescript, cleaned up tests. Rewrote a bunch of objects.

I'd love if you get a chance to try it before i cut a release on monday.

jdoose commented:
@halkeye , wow, that is an impressive piece of work!
I took a quick look over the code, looks good! And, yay, no coffeescript! :D

Unfortunately I won't be able to test it today - my family is visiting ;)

jdoose commented:
Hi @halkeye , I got it running in a docker environment with injected hubot-jenkins-notifier-script. (If you want it for testing purposes I am happy to send it to you!)

It worked for the STARTED phase but not for FINALIZED or FINISHED phase.

line
`JenkinsNotifierRequest.prototype.processFinished = JenkinsNotifierRequest.prototype.onFinalized = function(data) {`
should be changed to 
`JenkinsNotifierRequest.prototype.processFinished = JenkinsNotifierRequest.prototype.processFinalized = function(data) {`

line
`if (this.failed[data.name]) {` should be changed to `if (this.failing[data.name]) {`

'always_notify' is not respected.
I guess 'buildEnvelope' must be called before `process` (around line 244), right?

Anyway, this is an impressive piece of rework and way better to read and maintain! Thanks a lot!

halkeye commented:
okay, i'll take care of the 3 you mentioned.

Can you make a quick PR for to add more test cases for FINALIZED and FINISHED to commonBodies that you got from your script?

I'll take care of the rest.

jdoose commented:
Good morning Gavin,

sorry, but I don't know what you mean. I did not add any test cases. I set up a docker environment with a hubot installation that talks to a rocketchat in a docker container. A jenkins (also in a docker container) is set up to talk to the hubot with your current implementation, which I just copied to the hubot docker container.
The I did some manually testing with jenkins.

halkeye commented:
Okay well make it simple.

Do you still have the json from those manual tests? Otherwise I'll try and reproduce this week

jdoose commented:
Oh, sure! I can reproduce and just copy the dump output, no problem! I will do it later today when I'm home.

jdoose commented:
Build #69, will succeed
```
{ number: 69,
  queue_id: 69,
  phase: 'STARTED',
  url: 'job/Jobname/69/',
  scm: {},
  log: 'Started by user admin\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\n',
  artifacts: {} }
{ number: 69,
  queue_id: 69,
  phase: 'COMPLETED',
  status: 'SUCCESS',
  url: 'job/Jobname/69/',
  scm: {},
  log: 'Started by user admin\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\nBuilding in workspace /var/jenkins_home/workspace/Jobname\n[Jobname] $ /bin/sh -xe /tmp/hudson5174919973359476640.sh\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\n',
  artifacts: {} }
{ number: 69,
  queue_id: 69,
  phase: 'FINALIZED',
  status: 'SUCCESS',
  url: 'job/Jobname/69/',
  scm: {},
  log: 'Started by user admin\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\nBuilding in workspace /var/jenkins_home/workspace/Jobname\n[Jobname] $ /bin/sh -xe /tmp/hudson5174919973359476640.sh\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\nFinished: SUCCESS\n',
  artifacts: {} }
```

build #70, will fail:

```
{ number: 70,
  queue_id: 70,
  phase: 'STARTED',
  url: 'job/Jobname/70/',
  scm: {},
  log: 'Started by user admin\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\n',
  artifacts: {} }
{ number: 70,
  queue_id: 70,
  phase: 'COMPLETED',
  status: 'FAILURE',
  url: 'job/Jobname/70/',
  scm: {},
  log: 'Started by user admin\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\nBuilding in workspace /var/jenkins_home/workspace/Jobname\n[Jobname] $ /bin/sh -xe /tmp/hudson326640172349901710.sh\n+ fail and break this build!\n/tmp/hudson326640172349901710.sh: 3: /tmp/hudson326640172349901710.sh: fail: not found\nBuild step \'Execute shell\' marked build as failure\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\n',
  artifacts: {} }
{ number: 70,
  queue_id: 70,
  phase: 'FINALIZED',
  status: 'FAILURE',
  url: 'job/Jobname/70/',
  scm: {},
  log: 'Started by user admin\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\nBuilding in workspace /var/jenkins_home/workspace/Jobname\n[Jobname] $ /bin/sh -xe /tmp/hudson326640172349901710.sh\n+ fail and break this build!\n/tmp/hudson326640172349901710.sh: 3: /tmp/hudson326640172349901710.sh: fail: not found\nBuild step \'Execute shell\' marked build as failure\nNotifying endpoint \'HTTP:http://192.168.168.128:9001/hubot/jenkins-notify?room=general&always_notify=1&trace=1\'\nFinished: FAILURE\n',
  artifacts: {} }
```

Does that help?

halkeye commented:
buildEnvelope is called later in the request

```
        /* Send out all the messages */
        var envelope = JenkinsNotifierRequest.buildEnvelope(notifier.getQuery());
        lodash.forEach(messages, function(msg) {
          this.robot.send(envelope, msg);
        }.bind(this))
```

halkeye commented:
okay, fixed all the tests again

jdoose commented:
Hello!

What I meant about `buildEnvelope` is this:
First `process` is called, which will call `processFinished`.
After the completion of `process`, `buildEnvelope` will be called.

`processFinished` will use the notification strategy declared in `this.query.onFinished`.

But:
If just `always_notify`is specified then `buildEnvelope` will add the `onFinished` parameter to `envelope`. 
This will be too late since `processFinished` was already executed.

halkeye commented:
I'll have to check. build envelope is more for where its being sent, build
query string is the onFinished stuff

On Tue, Sep 13, 2016 at 1:41 AM, Jens Doose <notifications@github.com>
wrote:

> Hello!
>
> What I meant about buildEnvelope is this:
> First process is called, which will call processFinished.
> After the completion of process, buildEnvelope will be called.
>
> processFinished will use the notification strategy declared in
> this.query.onFinished.
>
> But:
> If just always_notifyis specified then buildEnvelope will add the
> onFinished parameter to envelope.
> This will be too late since processFinished was already executed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/halkeye/hubot-jenkins-notifier/pull/7#issuecomment-246614823>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGuB16SBoDIVwDnDJHy6zn8Iaqz9Sedks5qpmHBgaJpZM4JY0mC>
> .
>

halkeye commented:
oh you are totally right

I'll have to think of a test that catches this

halkeye commented:
Okay, updated.

Do you want to create a new ticket if you find anything else?
